### PR TITLE
[Templating] Deprecate the component

### PR DESCRIPTION
**Pull request by dunglas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | #15028
| License       | MIT
| Doc PR        | n/a

Follows #21035.

### CODE REVIEW

**Comment by dunglas** [comment]
So let's deprecate this notation.

**Comment by mickaelandrieu** [comment]
@dunglas are you sure ? The Templating component is still useful. I remember the core team wanted to remove it from framework, this didn't mean remove it from organization or deprecate it: am I wrong ?

**Comment by scaytrase** [comment]
@mickaelandrieu twig is first-level citizen for about two years already. I think the purpose is to remove unneccessary abstraction layer from templating to twig in the future for 4.0

http://symfony.com/blog/new-in-symfony-2-7-twig-as-a-first-class-citizen

**Comment by mickaelandrieu** [comment]
I agree to remove abstraction from standard edition, but I don't think we should deprecate the component as a component: there is life outside of full edition! And regarding the download stats we can already say that the components are more used outside the full edition :)

**Comment by wouterj** [comment]
Do we want to deprecate the PHP templating system or the Templating component? Those are 2 very different things. Imo, the templating component/abstraction still makes sense. The PHP templating might be deprecated though, as the general opinion in the community seems to be in favor of removing it.

**Comment by mickaelandrieu** [comment]
@wouterj 👍 yes, as we didn't use PHP templating anymore, it makes sense to remove every PHP template from FrameworkBundle.

I have a use case in favor of keeping Templating component: actually, PrestaShop project is using both Smarty and Twig engines.

For now, integrators can "only" use Smarty to make their own themes and plugins, but the idea is to migrate everything to Twig, if we can do it smoothly, ie without break all the things again. Note we are still using Symfony2, but this kind of deprecation can stop a future migration to Symfony3.

I guess PHPBB team may have the same needs too.

To sum up what I suggest:
* remove every use of Templating component from framework
* keep the component alive
* Remove/reword some cookbooks

Wdyt ?

**Comment by scaytrase** [comment]
I think keeping it alive (like it was done with `AsseticBundle` which is still useful after removing from standard) is good idea in any case

**Comment by nicolas-grekas** [comment]
Note that even deprecated, the component will be maintained until the end of 3.4LTS support, that means another 4 years from now.

**Comment by mneuhaus** [comment]
Just a +1 from a "bystander" for keeping general support for alternative templating engines, i'm currently working on bringing TYPO3's Fluid templating engine to Symfony :)

**Comment by dunglas** [comment]
@symfony/deciders Should we keep this component even if it isn't integrated in FrameworkBundle anymore or not?

**Comment by mickaelandrieu** [comment]
Components have a life outside the framework :)

Regarding the download statistics, the framework only represent *10℅* of the Symfony ecosystem.

Also, I've never seen such big effort to maintain this component: so why do you want to deprecate it?

We don't ask new features from Core team, only to keep up to date and report security issues if any :/

**Comment by dunglas** [comment]
As pointed out by @nicolas-grekas, this component will still be supported 4 years. If even Symfony itself don't use it anymore, and if it doesn't allow anymore to integrate other templating engines with Symfony I don't get the point of using it in a third party project.

**Comment by robfrawley** [comment]
Is this removing the ability to easily integrate other templating engines with Symfony? If so, I'm :-1:

**Comment by mickaelandrieu** [comment]
@dunglas it will be up to component' users to make the bridge with Symfony framework if they want to use it in a "framework context" if the core team doesn't want to support this feature anymore.

But if the component is deprecated, the right answer for "can we use PHP/Smarty/Blade/xXx in Symfony (framework)" will become "no" instead of "yes but need some work".

This is a really big difference from technical & business point of view!

Also, deprecate a component because it's not used in framework may be seen as a very bad news: should we expect a deprecation of every dependency removed from FrameworkBundle?

**Comment by fabpot** [comment]
@mickaelandrieu Stop the FUD :)

**Comment by mickaelandrieu** [comment]
really ? o_O meh do what you want, I'm done.

**Comment by javiereguiluz** [comment]
I like the idea of Symfony supporting only Twig out-of-the-box ... and I agree with deprecating this component.

However, just to know how to update the Symfony Docs, could you please tell me if it will be possible to use PHP/Smarty/Blade/... with Symfony after this change and a general overview of how to achieve that? Thanks!

**Comment by wouterj** [comment]
Well, as Symfony is 100% DIC based: Create a `smarty` service and do `$this->get('smarty')->render(...)` in the controller.

When this component is not deprecated, you would create a `SmartyEngine` and you can do `$this->render(...)` in the controller.

File: src/Symfony/Component/Templating/DelegatingEngine.php [hunk 1]
```diff
@@ -11,10 +11,14 @@
...
 namespace Symfony\Component\Templating;
 
+@trigger_error('The '.DelegatingEngine::class.' class is deprecated since version 3.3 and will be removed in 4.0. Use Twig instead.', E_USER_DEPRECATED);
+
 /**
  * DelegatingEngine selects an engine for a given template.
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @deprecated The DelegatingEngine class will be removed in Symfony 4.0. You should use Twig instead.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Templating/DelegatingEngine.php]: the annotations should be rewritten as 
`@deprecated since version 3.3, to be removed in 4.0. You should use Twig instead.`
this will make notices raised by DebugClassLoader a bit better/consistent.

**Comment by nicolas-grekas** [comment]
Status: needs work

**Comment by nicolas-grekas** [comment]
@fabpot can you share some hints about the direction you'd like this to take?

**Comment by fabpot** [comment]
I'm closing this one as I don't think there is a need to deprecate this component, just the integration with framework bundle, which is done in another PR.
